File: backend/app/schemas/sales_orders.py
# Request/response bodies for the sales orders module's endpoints.
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class CreateNewSalesOrderRequest(BaseModel):
    cust_id: int
    date: datetime
    # Parallel arrays, one entry per line item (mirrors
    # CreateNewPurchaseOrderRequest's product_ids/quantities/rates, plus
    # tax_percs since sales order tax is captured per line item rather than
    # as a flat order-level SGST/CGST/IGST amount).
    product_ids: list[int]
    quantities: list[int]
    rates: list[float]
    tax_percs: list[float]
    # Optional per-line remark, parallel to the arrays above — see
    # SalesSummary.note. A line without one sends ""; a client that omits
    # the whole list gets "" on every line.
    notes: list[str] | None = None
    # Flat discount off the order's whole net (pre-tax) amount — see
    # SalesOrders.overall_discount. Optional: an order without one submits 0.
    overall_discount: float = Field(default=0.0, ge=0)
    # Delivery charged to the customer, plus the GST % it carries — see
    # SalesOrders.delivery_charge. Both optional: an order with no delivery
    # submits 0 for each, which is exactly what an order raised before this
    # existed reads as.
    delivery_charge: float = Field(default=0.0, ge=0)
    delivery_tax_perc: float = Field(default=0.0, ge=0)
    description: str
    # Optional: purchase order(s) this sales order is fulfilled from.
    related_purchase_order_ids: list[int] = []
    # Optional: unbilled purchase order(s) it is fulfilled from — a separate
    # list because the two are different collections with overlapping ids.
    # See SalesOrders.related_unbilled_purchase_order_ids.
    related_unbilled_purchase_order_ids: list[int] = []

    @model_validator(mode="after")
    def _check_line_items_match(self) -> "CreateNewSalesOrderRequest":
        lengths = {len(self.product_ids), len(self.quantities), len(self.rates), len(self.tax_percs)}
        if self.notes is not None:
            lengths.add(len(self.notes))
        if len(lengths) != 1:
            raise ValueError(
                "product_ids, quantities, rates, tax_percs, and notes must have the same number of entries"
            )
        if len(self.product_ids) == 0:
            raise ValueError("at least one line item is required")
        return self


class CreateNewSalesOrderResponse(BaseModel):
    message: str


class SalesOrderDetailItem(BaseModel):
    id: int
    order_no: int
    order_status_id: int
    cust_id: int
    date: datetime
    # SalesSummary.id of each line, parallel to the arrays below. The edit
    # form sends these back as line_item_ids so an edit updates the same rows
    # in place — the "Add details" costing hangs off them (see
    # models/sales_order_costing.py).
    line_item_ids: list[int]
    product_ids: list[int]
    quantities: list[int]
    rates: list[float]
    tax_percs: list[float]
    # Per-line remark, "" where none was left — see SalesSummary.note.
    notes: list[str]
    overall_discount: float
    delivery_charge: float
    delivery_tax_perc: float
    total_amount_before_tax: float
    total_tax_amount: float
    total_amount_after_tax: float
    description: str
    related_purchase_order_ids: list[int]
    related_unbilled_purchase_order_ids: list[int] = []
    # See SalesOrders.po_updated_flag — true when a related purchase order
    # was edited since this sales order was last saved.
    po_updated_flag: bool
    is_deleted: bool


class UpdateSalesOrderDetailsRequest(BaseModel):
    id: int
    # order_status_id and is_deleted are only ever submitted on update — a
    # new sales order is always created as "New" and not deleted.
    order_status_id: int
    is_deleted: bool = False
    cust_id: int
    date: datetime
    # Parallel to product_ids: the SalesSummary.id each line came from (as
    # returned by get_sales_order_details), or None for a line added on the
    # form. Lets the edit update existing rows in place so their ids — and
    # the costing keyed on them — survive; a row whose id is not sent back is
    # deleted. Optional: a client that omits the whole list gets a
    # best-effort positional match instead (see update_sales_order_details).
    line_item_ids: list[int | None] | None = None
    product_ids: list[int]
    quantities: list[int]
    rates: list[float]
    tax_percs: list[float]
    # Optional per-line remark, parallel to the arrays above — see
    # SalesSummary.note. A line without one sends ""; a client that omits
    # the whole list gets "" on every line.
    notes: list[str] | None = None
    # Flat discount off the order's whole net (pre-tax) amount — see
    # SalesOrders.overall_discount. Optional: an order without one submits 0.
    overall_discount: float = Field(default=0.0, ge=0)
    # Delivery charged to the customer, plus the GST % it carries — see
    # SalesOrders.delivery_charge. Both optional: an order with no delivery
    # submits 0 for each, which is exactly what an order raised before this
    # existed reads as.
    delivery_charge: float = Field(default=0.0, ge=0)
    delivery_tax_perc: float = Field(default=0.0, ge=0)
    description: str
    related_purchase_order_ids: list[int] = []
    related_unbilled_purchase_order_ids: list[int] = []

    @model_validator(mode="after")
    def _check_line_items_match(self) -> "UpdateSalesOrderDetailsRequest":
        lengths = {len(self.product_ids), len(self.quantities), len(self.rates), len(self.tax_percs)}
        if self.line_item_ids is not None:
            lengths.add(len(self.line_item_ids))
        if self.notes is not None:
            lengths.add(len(self.notes))
        if len(lengths) != 1:
            raise ValueError(
                "line_item_ids, product_ids, quantities, rates, tax_percs, and notes "
                "must have the same number of entries"
            )
        if len(self.product_ids) == 0:
            raise ValueError("at least one line item is required")
        if self.line_item_ids is not None:
            sent_ids = [line_item_id for line_item_id in self.line_item_ids if line_item_id is not None]
            if len(sent_ids) != len(set(sent_ids)):
                raise ValueError("each line item id may only appear once")
        return self


class UpdateSalesOrderDetailsResponse(BaseModel):
    message: str


# The sales orders table's inline status dropdown. Deliberately NOT the full
# UpdateSalesOrderDetailsRequest with one field swapped: that endpoint
# rewrites #sales_summary, re-derives the totals, and clears po_updated_flag,
# none of which an admin flipping a status in a table row is asking for.
class UpdateSalesOrderStatusRequest(BaseModel):
    id: int
    order_status_id: int


class UpdateSalesOrderStatusResponse(BaseModel):
    message: str


class OrderStatusListItem(BaseModel):
    id: int
    status_name: str

File: frontend/src/lib/sales-orders.ts
// ---------------------------------------------------------------------------
// Sales order data for the /admin/orders "Sales orders" tab
// ---------------------------------------------------------------------------
// Mirrors lib/purchase-orders.ts. Fetches from GET /admin/get_sales_order_details
// (backend/app/api/routes/sales_orders.py), which only ever returns active
// orders — soft-deleted ones are filtered out server-side so they can't be
// viewed.
//
// SalesOrders itself has no line-item field — product/quantity/rate/tax rows
// belong to the separate #sales_summary collection (backend/app/models/
// sales_summary.py), linked back via its sales_order_id FK.
// get_sales_order_details folds those rows back in as parallel
// productIds/quantities/rates/taxPercs arrays, same convention as
// lib/purchase-orders.ts. sales-order-form-modal.tsx submits/edits them the
// same way, as parallel product_ids/quantities/rates/tax_percs arrays, to
// create_new_sales_order/update_sales_order_details.
//
// order_no and order_status_id are backend-assigned on create (order_no via
// OrderNoCounterMaster, order_status_id defaulted to the seeded "New"
// row) — the create request has neither field; only update_sales_order_details
// accepts order_status_id.
import { apiFetch } from "@/lib/api";

export type SalesOrder = {
  id: number;
  orderNo: number;
  orderStatusId: number;
  custId: number;
  date: string;
  productIds: number[];
  quantities: number[];
  rates: number[];
  taxPercs: number[];
  // Flat discount off the whole order's net (pre-tax) amount, entered on the
  // order form. It is already baked into the three totals below — the
  // backend splits it across the line items before charging tax (see
  // _allocate_overall_discount in backend/app/api/routes/sales_orders.py) —
  // so nothing here should ever subtract it a second time.
  overallDiscount: number;
  // Delivery billed to the customer, and the GST % it carries. Also already
  // inside the three totals below — the backend adds it on top of the line
  // items and taxes it in its own right (see SalesOrders.delivery_charge).
  // Both 0 on an order with no delivery, and on every order raised before
  // the field existed.
  deliveryCharge: number;
  deliveryTaxPerc: number;
  totalAmountBeforeTax: number;
  totalTaxAmount: number;
  totalAmountAfterTax: number;
  description: string;
  relatedPurchaseOrderIds: number[];
  // Unbilled purchase orders this sales order is fulfilled from — a separate
  // list because the two live in different collections with overlapping ids.
  // See backend/app/models/sales_orders.py.
  relatedUnbilledPurchaseOrderIds: number[];
  // True once a related purchase order (on EITHER list above) has been
  // edited since this sales order was last saved — a review notice, not an
  // automatic data sync; see backend/app/models/sales_orders.py's
  // po_updated_flag docstring. Cleared server-side the next time this sales
  // order is saved via update_sales_order_details.
  poUpdatedFlag: boolean;
  isDeleted: boolean;
};

// Shape returned by the backend's SalesOrderDetailItem schema.
type SalesOrderDetailItem = {
  id: number;
  order_no: number;
  order_status_id: number;
  cust_id: number;
  date: string;
  product_ids: number[];
  quantities: number[];
  rates: number[];
  tax_percs: number[];
  overall_discount: number;
  delivery_charge: number;
  delivery_tax_perc: number;
  total_amount_before_tax: number;
  total_tax_amount: number;
  total_amount_after_tax: number;
  description: string;
  related_purchase_order_ids: number[];
  related_unbilled_purchase_order_ids: number[];
  po_updated_flag: boolean;
  is_deleted: boolean;
};

function toSalesOrder(item: SalesOrderDetailItem): SalesOrder {
  return {
    id: item.id,
    orderNo: item.order_no,
    orderStatusId: item.order_status_id,
    custId: item.cust_id,
    date: item.date,
    productIds: item.product_ids,
    quantities: item.quantities,
    rates: item.rates,
    taxPercs: item.tax_percs,
    // ?? 0 for orders raised before order-level discounts existed.
    overallDiscount: item.overall_discount ?? 0,
    // ?? 0 for orders raised before delivery charges existed.
    deliveryCharge: item.delivery_charge ?? 0,
    deliveryTaxPerc: item.delivery_tax_perc ?? 0,
    totalAmountBeforeTax: item.total_amount_before_tax,
    totalTaxAmount: item.total_tax_amount,
    totalAmountAfterTax: item.total_amount_after_tax,
    description: item.description,
    relatedPurchaseOrderIds: item.related_purchase_order_ids,
    relatedUnbilledPurchaseOrderIds: item.related_unbilled_purchase_order_ids ?? [],
    poUpdatedFlag: item.po_updated_flag,
    isDeleted: item.is_deleted,
  };
}

export async function fetchSalesOrders(): Promise<SalesOrder[]> {
  const response = await apiFetch("/admin/get_sales_order_details");
  if (!response.ok) {
    throw new Error("Failed to load sales orders");
  }

  const items: SalesOrderDetailItem[] = await response.json();
  return items.map(toSalesOrder);
}

// Backs the status dropdown in each row of the sales orders table. Its own
// endpoint rather than a full update_sales_order_details round trip: that
// one re-saves the whole order (rewriting the line items, recomputing the
// totals, clearing poUpdatedFlag), none of which changing a status should do.
//
// It can legitimately fail — moving an order into "Delivered"/"Completed"
// takes its quantities out of stock, and the backend rejects the move when
// there isn't enough — so the caller gets the reason back rather than a
// boolean. Resolves to null on success, or the message to show on failure.
export async function updateSalesOrderStatus(id: number, orderStatusId: number): Promise<string | null> {
  try {
    const response = await apiFetch("/admin/update_sales_order_status", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ id, order_status_id: orderStatusId }),
    });

    if (response.ok) {
      return null;
    }

    // Surface the backend's actual reason (e.g. "insufficient stock for:
    // product 12 (on hand 3, needs 2 more)") instead of guessing.
    const body = await response.json().catch(() => null);
    return typeof body?.detail === "string" ? body.detail : "Couldn't update the status. Please try again.";
  } catch {
    return "Couldn't reach the server. Please try again.";
  }
}
